File: TimeConverter.js
/* Time Converter
 * This is open source, please refer to the license
 * for details.
 * Time Converter converts Modern Western to Classical Chinese
 * systems of time measurement. Future versions may do the reverse, or allow
 * more generalized methods for converting to other 
 * non-standard time systems.
 * Time converter gives more accurate time information, down to the
 * millisecond. And access to the JS Date function for greater utility.
 */

/* Conversion chart for millisenconds:
 * 1 Miao =         144 ms (100 per Fen)
 * 1 Fen =        14400 ms (500 per Shi, 60 per Ke)
 * 1 XiaoKe =    144000 ms (50 per Shi)
 * 1 Ke =        864000 ms (100 per Day)
 * 1 ShiChen =  7200000 ms (12 per Day)
 * 1 Day =     86400000 ms
 * Xun are varriable, some have 10 and some have 9 days.
 * 1 Xun =    864000000 ms -or-
 *            777600000 ms (3 per month)
 * Months are variable, some have 30 and some have 29 days.
 * 1 Month = 2592000000 ms -or-
 *           2505600000 ms (12 or 13 per year)
 * Years are varriable, some have leap months added to them.
 * 1 Year = 12 or 13 Months
 * 1 Sixtyear = 60 years
 * 1 Great Cycle = 3 sixtyears
 * The first day of the first year of the first Great Cycle was in 2697 BC.
 * CCT recognizes time zones, but not daylight savings time.
 */

//public data table
var tmprlRA = [];

//the core converter. Uses ms because that's the base time unit in Date().
function convToCCT(ms) { 
    var rawMiao = 0,
	miao = 0,
	fen = 0,
	xiaoKe = 0,
	ke = 0,
	shiChenNum = 0,
	day = 0,
	localRA = [];

    rawMiao = ms / 144;
    miao = Math.floor(rawMiao) % 100;

    if (rawMiao >= 100) {
	fen = Math.floor(rawMiao / 100) % 60;
    } if (rawMiao >= 1000) {
	xiaoKe = Math.floor(rawMiao / 1000) % 50;
    } if (rawMiao >= 6000) {
	ke = Math.floor(rawMiao / 6000) % 100;
    } if (rawMiao >= 50000) {
	shiChenNum = Math.floor(rawMiao / 50000) % 12;
    } if (rawMiao >= 600000) {
	day = Math.floor(rawMiao / 600000);
    }

    localRA.push(day, shiChenNum, ke, xiaoKe, fen, miao);
    tmprlRA = localRA;
    return tmprlRA;
}

//separating out display from calc so it's easy to delete.
function displayCCT (ms) {
    convToCCT(ms);

    var shiChen = "";

    switch(tmprlRA[1]) {
	case 0:
	    shiChen = "Zi";
	    break;
	case 1:
	    shiChen = "Chou";
	    break;
	case 2:
	    shiChen = "Yin";
	    break;
	case 3:
	    shiChen = "Mao";
	    break;
	case 4:
	    shiChen = "Chen";
	    break;
	case 5:
	    shiChen = "Si";
	    break;
	case 6:
	    shiChen = "Wu";
	    break;
	case 7:
	    shiChen = "Wei";
	    break;
	case 8:
	    shiChen = "Shen";
	    break;
	case 9:
	    shiChen = "You";
	    break;
	case 10:
	    shiChen = "Xu";
	    break;
	case 11:
	    shiChen = "Hai";
	    break;
    }

    //stringifies numbers after the Shi place (tmprlRA[1]) if < 10
    //so they always have an even two spaces.
    for (var i = 2; i < tmprlRA.length; i++) {
	if (tmprlRA[i] < 10 ) {
	    tmprlRA[i].toString;
	    tmprlRA[i] = "0" + tmprlRA[i]; 
	}
    }

    var display = "Day " + tmprlRA[0] + ". " 
		+ shiChen + " Shi " + tmprlRA[3] + ". " 
		+ "Ke " + tmprlRA[2] + ":" + tmprlRA[4] + ":" 
		+ tmprlRA[5] + " o'day.";

    return display;
}

function getCCTNow(TZ) {
    var nowUTC = new Date,
	nowUTCms = Date.parse(nowUTC),
	nowTZ = 0,
	minute = 60000,
	hour = 3600000;
	//the hour earlier is because Classical Chinese time 
	//recons the start of day at 2300 hrs.
	if (TZ === undefined ) {
		nowTZ = -(nowUTC.getTimezoneOffset() * minute - hour);
	} else { 
		nowTZ = TZ * hour + hour;
	}

	now = nowUTCms + nowTZ;

    return displayCCT(now);
}

//maual tests:
console.log(getCCTNow(-8));
console.log(Date());

File: TimeConverter.py
# python port of TimeConverter.js
# Inclomplete, OBVS

def convToCCT(ms,tmprlRA):
    rawMiao = ms // 144
    miao = rawMiao % 100
    fen = 0
    xiaoKe = 0
    ke = 0
    shiChenNum = 0
    day = 0
    localRA = [];

    #Can all this be part of the var's definitions?
    if rawMiao >= 100:
	fen = (rawMiao // 100) % 60
    elif rawMiao >= 1000:
	xiaoKe = (rawMiao // 1000) % 50
    elif rawMiao >= 6000:
	ke = (rawMiao // 6000) % 100;
    elif rawMiao >= 50000:
	shiChenNum = (rawMiao // 50000) % 100
    elif rawMiao >= 600000:
	day = (rawMiao // 600000)

    #are both of these steps necessary?
    localRA.append(day)
    localRA.append(shiChenNum)
    localRA.append(ke)
    localRA.append(xiaoKe)
    localRA.append(fen)
    localRA.append(miao)

    tmprlRA = localRA;

def displayCCT (ms,tmprlRA):
    convToCCT(ms,tmprlRA)

    #Thows error: List Index out of range.
	#At the risk of sounding ignorant, 
	#why can't you just access it in python?
    #Option 1: Convert tmprlRA to a Dictionary. I know that works.
    #Option 2: find the sintax that allows list access.
    num = tmprlRA[1]
    shiChen = ''

    if num == 0: shiChen = 'Zi'
    elif num == 1: shiChen = 'Chou'
    elif num == 2: shiChen = 'Yin'
    elif num == 3: shiChen = 'Mao'
    elif num == 4: shiChen = 'Chen'
    elif num == 5: shiChen = 'Si'
    elif num == 6: shiChen = 'Wu'
    elif num == 7: shiChen = 'Wei'
    elif num == 8: shiChen = 'Shen'
    elif num == 9: shiChen = 'You'
    elif num == 10: ShiChen = 'Xu'
    else: #if num == 11:
	shiChen = 'Hai'

    for i in tmprlRA:
	if tmprlRA[i] < 10:
	    str(tmprlRA[i])
	    tmprlRA[i] = '0' + tmprlRA[i]

    display = 'day %s. %s Shi %s. Ke %s:%s:%s o\'day.' % (tmprlRA[1],tmprlRA[2],tmprlRA[3],tmprlRA[4],tmprlRA[5])

    return display

def main():
    tmprlRA = []
    print displayCCT(1465677620491,tmprlRA)
    #Close to: Day 16963. Wu Shi 41. Ke 56:53:05 o'day.
if __name__ == '__main__':
    main()
